File: laba5/ex3/tram_manager.cpp
// tram_manager.cpp
#include "tram_manager.h"
#include <algorithm>
#include <set>

std::string TramManager::createTram(const std::string& name, const std::vector<std::string>& stops) {
    // Проверка на минимальное количество остановок
    if (stops.size() < 2) {
        return "ERROR: At least two stops required";
    }
    
    // Проверка на повторяющиеся остановки
    std::set<std::string> unique_stops(stops.begin(), stops.end());
    if (unique_stops.size() != stops.size()) {
        return "ERROR: Duplicate stops in route";
    }
    
    // Проверка на существующее имя
    if (tram_routes_.count(name)) {
        return "ERROR: Tram '" + name + "' already exists";
    }
    
    // Проверка на дубликат маршрута
    for (const auto& [existing_name, existing_route] : tram_routes_) {
        if (existing_route == stops) {
            return "ERROR: Route duplicates existing tram '" + existing_name + "'";
        }
    }
    
    // Добавление нового маршрута
    tram_routes_[name] = stops;
    for (const auto& stop : stops) {
        stop_trams_[stop].insert(name);
    }
    
    return "";
}

// Остальные методы остаются без изменений
std::vector<std::string> TramManager::getTramsInStop(const std::string& stop) const {
    auto it = stop_trams_.find(stop);
    return (it != stop_trams_.end()) 
        ? std::vector<std::string>(it->second.begin(), it->second.end())
        : std::vector<std::string>();
}

std::vector<std::pair<std::string, std::set<std::string>>> TramManager::getStopsInTram(const std::string& tram) const {
    std::vector<std::pair<std::string, std::set<std::string>>> result;
    auto tram_it = tram_routes_.find(tram);
    
    if (tram_it != tram_routes_.end()) {
        for (const auto& stop : tram_it->second) {
            std::set<std::string> other_trams;
            if (auto stop_it = stop_trams_.find(stop); stop_it != stop_trams_.end()) {
                std::copy_if(stop_it->second.begin(), stop_it->second.end(),
                            std::inserter(other_trams, other_trams.end()),
                            [&tram](const auto& name) { return name != tram; });
            }
            result.emplace_back(stop, other_trams);
        }
    }
    return result;
}

std::map<std::string, std::vector<std::string>> TramManager::getAllTrams() const {
    return tram_routes_;
}

File: laba5/ex3go/tram_manager.go
package main

import (
	"errors"
	"fmt"
	"sort"
)

type TramManager struct {
	tramRoutes map[string][]string    // Маршруты трамваев: имя → остановки
	stopTrams  map[string]map[string]struct{} // Остановки → трамваи
}

func NewTramManager() *TramManager {
	return &TramManager{
		tramRoutes: make(map[string][]string),
		stopTrams:  make(map[string]map[string]struct{}),
	}
}

func (tm *TramManager) CreateTram(name string, stops []string) error {
	// Проверка на минимальное количество остановок
	if len(stops) < 2 {
		return errors.New("ERROR: At least two stops required")
	}

	// Проверка на повторяющиеся остановки
	uniqueStops := make(map[string]struct{})
	for _, stop := range stops {
		if _, exists := uniqueStops[stop]; exists {
			return errors.New("ERROR: Duplicate stops in route")
		}
		uniqueStops[stop] = struct{}{}
	}

	// Проверка на существующее имя
	if _, exists := tm.tramRoutes[name]; exists {
		return fmt.Errorf("ERROR: Tram '%s' already exists", name)
	}

	// Проверка на дубликат маршрута
	for existingName, existingRoute := range tm.tramRoutes {
		if slicesEqual(existingRoute, stops) {
			return fmt.Errorf("ERROR: Route duplicates existing tram '%s'", existingName)
		}
	}

	// Добавление нового маршрута
	tm.tramRoutes[name] = stops
	for _, stop := range stops {
		if tm.stopTrams[stop] == nil {
			tm.stopTrams[stop] = make(map[string]struct{})
		}
		tm.stopTrams[stop][name] = struct{}{}
	}

	return nil
}

func (tm *TramManager) GetTramsInStop(stop string) []string {
	trams, exists := tm.stopTrams[stop]
	if !exists {
		return nil
	}

	result := make([]string, 0, len(trams))
	for tram := range trams {
		result = append(result, tram)
	}
	sort.Strings(result)
	return result
}

func (tm *TramManager) GetStopsInTram(tram string) map[string][]string {
	stops, exists := tm.tramRoutes[tram]
	if !exists {
		return nil
	}

	result := make(map[string][]string)
	for _, stop := range stops {
		trams := make([]string, 0)
		for t := range tm.stopTrams[stop] {
			if t != tram {
				trams = append(trams, t)
			}
		}
		sort.Strings(trams)
		result[stop] = trams
	}

	return result
}

func (tm *TramManager) GetAllTrams() map[string][]string {
	return tm.tramRoutes
}

func slicesEqual(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
